pylint/checkers/variables.py

```diff
@@ -665,6 +673,149 @@ def get_next_to_consume(self, node: nodes.Name) -> list[nodes.NodeNG] | None:
+
+    @staticmethod
+    def _branch_handles_name(name: str, body: Iterable[nodes.NodeNG]) -> bool:
+        return any(
+            NamesConsumer._defines_name_raises_or_returns(name, if_body_stmt)
+            or isinstance(
+                if_body_stmt,
+                (nodes.If, nodes.TryExcept, nodes.TryFinally, nodes.With),
+            )
+            and NamesConsumer._exhaustively_define_name_raise_or_return(
+                name, if_body_stmt
+            )
+            for if_body_stmt in body
+        )
+
+    def _uncertain_nodes_in_false_tests(
+        self, found_nodes: list[nodes.NodeNG], node: nodes.NodeNG
+    ) -> list[nodes.NodeNG]:
+        """Identify nodes of uncertain execution because they are defined under
+        tests that evaluate false.
+
+        Don't identify a node if there is a collectively exhaustive set of paths
+        that define the name, raise, or return (e.g. every if/else branch).
+        """
+        uncertain_nodes = []
+        for other_node in found_nodes:
+            if in_type_checking_block(other_node):
+                continue
+
+            if not isinstance(other_node, nodes.AssignName):
+                continue
+
+            closest_if = utils.get_node_first_ancestor_of_type(other_node, nodes.If)
+            if closest_if is None:
+                continue
+            if node.frame() is not closest_if.frame():
+                continue
+            if closest_if is not None and closest_if.parent_of(node):
+                continue
+            if any(
+                if_node.parent_of(closest_if)
+                for if_node in self._if_nodes_deemed_uncertain
+            ):
+                uncertain_nodes.append(other_node)
+                continue
+
+            # All inferred values must test false
+            if isinstance(closest_if.test, nodes.NamedExpr):
+                test = closest_if.test.value
+            else:
+                test = closest_if.test
+            all_inferred = utils.infer_all(test)
+            if not all_inferred or not all(
+                isinstance(inferred, nodes.Const) and not inferred.value
+                for inferred in all_inferred
+            ):
+                continue
+
+            # Name defined in every if/else branch
+            if NamesConsumer._exhaustively_define_name_raise_or_return(
+                other_node.name, closest_if
+            ):
+                continue
+
+            # Name defined somewhere else during a loop
+            if NamesConsumer._name_defined_later_in_loop(closest_if, other_node):
+                continue
+
+            uncertain_nodes.append(other_node)
+            self._if_nodes_deemed_uncertain.add(closest_if)
+
+        return uncertain_nodes
+
+    @staticmethod
+    def _name_defined_later_in_loop(
+        closest_if: nodes.If, other_node: nodes.NodeNG
+    ) -> bool:
+        """Identify if the name tested in `closest_if` is assigned later in a loop."""
```

Comment:
Let's use inference if it can simplify things. It's "costly" in term of performance but not if we have time to make it better in astroid, and it's a lot easier to maintain than something clever in pylint so it could be a virtuous cycle where it create time to make astroid faster.

Reply:
(Unrelated but this is what I have in mind after reviewing a lot of issues recentely and half of them are control flow issue : It's the same thing for control flow, if we can have a clean generic way to do it in astroid, it could avoid dozen of little checks in pylint.)